### DISCUSSION TITLE
Synchronous path definitions, global shared data and thread safety

### DISCUSSION THREAD

**Discussion by edward-b-1** [fastapi/fastapi#11878] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# pseudocode

some_global_shared_data = 0

@app.post('/api/increment')
def increment_global_shared_data():
    some_global_shared_data += 1

@app.get('/api/get_value')
def get_value_of_global_shared_data():
    return {
        'current_value': some_global_shared_data,
    }
```

### Description

I came across something in the documentation which suddenly triggered an alarm bell in my mind.

If I have understood the documentation correctly, `async def` paths are run in a single thread as part of an event loop which handles asynchronous tasks. Since there is only one thread involved, race conditions are not a consideration.

However, what I did not realize is that synchronous `def` paths are run by a thread as part of a thread pool. (Is the default number of threads 40?)

This is when I started to become concerned. My application contains synchronous `def` endpoints, and these endpoints access and modify global shared data. The pseudocode I have added is enough to show the structure of the application.

There isn't anything in the documentation about this. That could be for a number of reasons:

- FastAPI is not expected to be used in this way, with global, mutable, shared data. (In other words, creating a global mutable shared data is wrong/disallowed.)
- There is another convenient way to manage global shared mutable data with FastAPI framework. (I am not aware of it)
- The author of the documentation assumed this would be an obvious problem when using synchronous `def` paths, and that global mutable shared data should be protected by a mutex or lock, so didn't mention it explicitly
- Or: the current documentation is missing something like a note warning users not to create global, mutable state when defining `def` endpoints

Could anyone comment on this or help to clarify how FastAPI works in this context and whether this issue of global mutable state really is a concern if writing synchronous paths, or if it is not a concern and I have misunderstood something?

### Operating System

Linux

### FastAPI Version

0.111.1

### Pydantic Version

2.8.2

### Python Version

3.11.2

**Comment by YuriiMotov** [comment]
> Since there is only one thread involved, race conditions are not a consideration.

It's wrong.. Take a look at the pseudo-code below:

```
some_global_shared_data = 0

@app.post('/api/operation')
async def do_operation():
    my_var = some_global_shared_data
    res = await some_function(my_var)
    some_global_shared_data = res
```

Every time you use `await` it let's python switch to another async task.
So you have to be careful regardless of whether you use sync or async endpoints.

**Reply by edward-b-1** [reply]
To clarify, I was assuming that `await` is not called in any `async def` endpoint.

**Comment by YuriiMotov** [comment]
I think this question is out of scope of FastAPI docs.
If we try to cover every edge case we will end up having huge and unreadable docs.

So, let's close this. And this discussion will probably help those people who face such situation.
Thanks for raising this question!
If you disagree with my opinion, feel free to suggest changes to docs and open PR
